File: cloudflare/alife-app/src/utils/eventCreationArrangements.ts
import { localTimeToUtc } from '../../../shared/eventDetails.ts'
import type { EventActivityType, EventPlanProposal, LocalizedText } from '../types/eventComposition'
import type { CreationDraft } from './eventCreationDraft'

export type TimedArrangement = { id: string; startLocal: string; endLocal: string }
export type CreationSlot = TimedArrangement & { roleCode: string; requiredCount: string; eligibilityCode: string }
export type CreationProgramItem = { id: string; title: LocalizedText; description: LocalizedText; startOffsetMinutes: string; durationMinutes: string }
export type CreationSession = TimedArrangement & { title: LocalizedText; items: CreationProgramItem[] }
export type CreationVenue = TimedArrangement & {
  venueId: string; venueETag: string; name: LocalizedText; address: LocalizedText; capacity: string; requiredCapacity: string
}
export type CreationArrangements = { slots?: CreationSlot[]; sessions?: CreationSession[]; venues?: CreationVenue[] }
export type EventCreationArrangementsRequest = {
  serviceSlots?: { roleCode: string; requiredCount: number; eligibilityCode: string; startOffsetMinutes: number; endOffsetMinutes: number }[]
  sessions?: { title: LocalizedText; startOffsetMinutes: number; endOffsetMinutes: number; items: { title: LocalizedText; description: LocalizedText; startOffsetMinutes: number; durationMinutes: number }[] }[]
  venueBookings?: { venueId: string | null; venueETag: string | null; newVenue: { name: LocalizedText; address: LocalizedText; capacity: number; isActive: boolean } | null; requiredCapacity: number; startOffsetMinutes: number; endOffsetMinutes: number }[]
}
export const arrangementEnabled = (proposal: EventPlanProposal, code: string) => proposal.moduleDecisions.some(item => item.moduleCode === code && item.status !== 'inactive')
export const creationSlots = (draft: CreationDraft, type: EventActivityType): CreationSlot[] => draft.arrangements?.slots ?? type.presetServiceSlots.map(slot => ({
  id: `preset-${slot.roleCode}`, roleCode: slot.roleCode, requiredCount: String(slot.requiredCount), eligibilityCode: slot.eligibilityCode,
  startLocal: draft.startLocal, endLocal: draft.endLocal,
}))
export const arrangementTiming = (draft: CreationDraft, row: TimedArrangement) => {
  const base = Date.parse(localTimeToUtc(draft.startLocal, draft.timeZone))
  return {
    startOffsetMinutes: (Date.parse(localTimeToUtc(row.startLocal, draft.timeZone)) - base) / 60000,
    endOffsetMinutes: (Date.parse(localTimeToUtc(row.endLocal, draft.timeZone)) - base) / 60000,
  }
}

export function creationArrangements(draft: CreationDraft, type: EventActivityType, proposal: EventPlanProposal): EventCreationArrangementsRequest {
  return {
    serviceSlots: arrangementEnabled(proposal, 'SERVICE.ROSTER') ? creationSlots(draft, type).map(slot => ({
      roleCode: slot.roleCode.trim(), requiredCount: Number(slot.requiredCount), eligibilityCode: slot.eligibilityCode, ...arrangementTiming(draft, slot),
    })) : [],
    sessions: arrangementEnabled(proposal, 'PROGRAM.PRODUCTION') ? (draft.arrangements?.sessions ?? []).map(session => ({
      title: session.title, ...arrangementTiming(draft, session), items: session.items.map(item => ({
        title: item.title, description: item.description, startOffsetMinutes: Number(item.startOffsetMinutes), durationMinutes: Number(item.durationMinutes),
      })),
    })) : [],
    venueBookings: arrangementEnabled(proposal, 'PLACE.RESOURCE') ? (draft.arrangements?.venues ?? []).map(venue => ({
      venueId: venue.venueId || null, venueETag: venue.venueETag || null,
      newVenue: venue.venueId ? null : { name: venue.name, address: venue.address, capacity: Number(venue.capacity), isActive: true },
      requiredCapacity: Number(venue.requiredCapacity), ...arrangementTiming(draft, venue),
    })) : [],
  }
}

const bilingual = (value: LocalizedText, limit: number) => Boolean(value.en.trim() && value.zh.trim() && value.en.length <= limit && value.zh.length <= limit)
const positive = (value: string, max: number) => /^\d+$/.test(value) && Number(value) >= 1 && Number(value) <= max
export function validateCreationArrangements(draft: CreationDraft, type: EventActivityType, proposal: EventPlanProposal, zh: boolean): string {
  const validTime = (row: TimedArrangement) => {
    try {
      const timing = arrangementTiming(draft, row)
      return Number.isInteger(timing.startOffsetMinutes) && Number.isInteger(timing.endOffsetMinutes) && timing.startOffsetMinutes >= -10080 && timing.endOffsetMinutes <= 44640 && timing.endOffsetMinutes > timing.startOffsetMinutes
    } catch { return false }
  }
  if (arrangementEnabled(proposal, 'SERVICE.ROSTER')) {
    const slots = creationSlots(draft, type)
    if (!slots.length || slots.length > 50 || slots.some(slot => !slot.roleCode.trim() || slot.roleCode.length > 100 || !positive(slot.requiredCount, 10000) || !validTime(slot)))
      return zh ? '请展开「岗位与轮班」，填写岗位、人数及有效起止时间（最多 50 个岗位）。' : 'Open Roles and shifts and enter roles, counts and valid times (up to 50 slots).'
  }
  if (arrangementEnabled(proposal, 'PROGRAM.PRODUCTION')) {
    const sessions = draft.arrangements?.sessions ?? []
    if (!sessions.length || sessions.length > 20 || sessions.some(session => !bilingual(session.title, 240) || !validTime(session) || !session.items.length || session.items.length > 50))
      return zh ? '请展开「节目与制作」，填写双语环节名称、有效时间和至少一个节目。' : 'Open Programme and production and enter bilingual session titles, valid times and at least one programme item.'
    if (sessions.some(session => session.items.some(item => !bilingual(item.title, 240) || !/^\d+$/.test(item.startOffsetMinutes) || !positive(item.durationMinutes, 54720) || item.description.en.length > 2000 || item.description.zh.length > 2000 || Number(item.startOffsetMinutes) + Number(item.durationMinutes) > arrangementTiming(draft, session).endOffsetMinutes - arrangementTiming(draft, session).startOffsetMinutes)))
      return zh ? '请检查节目双语名称、开始分钟和时长；节目须在所属环节时间内。' : 'Check bilingual programme titles, offsets and durations. Each item must fit inside its session.'
  }
  if (arrangementEnabled(proposal, 'PLACE.RESOURCE')) {
    const venues = draft.arrangements?.venues ?? []
    if (!venues.length || venues.length > 20 || venues.some(venue => !validTime(venue) || !positive(venue.requiredCapacity, 1000000) || !positive(venue.capacity, 1000000) || Number(venue.requiredCapacity) > Number(venue.capacity) || (venue.venueId ? !venue.venueETag : !bilingual(venue.name, 240) || venue.address.en.length > 1000 || venue.address.zh.length > 1000)))
      return zh ? '请展开「场地与资源」，选择或填写场地、容量和预订时间；所需人数不可超过场地容量。' : 'Open Venue and resources and choose or enter venues, capacity and booking times. Attendance must fit the venue capacity.'
    for (let i = 0; i < venues.length; i++) for (let j = i + 1; j < venues.length; j++) {
      const a = venues[i], b = venues[j]
      if (a.venueId && a.venueId === b.venueId) {
        const x = arrangementTiming(draft, a), y = arrangementTiming(draft, b)
        if (x.startOffsetMinutes < y.endOffsetMinutes && y.startOffsetMinutes < x.endOffsetMinutes)
          return zh ? '同一场地的预订时间不能重叠，请在活动安排中调整。' : 'Bookings for the same venue cannot overlap. Adjust them in Arrangements.'
      }
    }
  }
  return ''
}

export function validStoredArrangements(value: unknown): value is CreationArrangements {
  if (!value || typeof value !== 'object' || Array.isArray(value)) return false
  const data = value as CreationArrangements
  const strings = (row: object, keys: string[]) => keys.every(key => typeof (row as Record<string, unknown>)[key] === 'string')
  const text = (x: unknown) => Boolean(x && typeof x === 'object' && strings(x, ['en', 'zh']))
  const timed = (x: TimedArrangement) => Boolean(x && strings(x, ['id', 'startLocal', 'endLocal']))
  return (data.slots === undefined || Array.isArray(data.slots) && data.slots.length <= 50 && data.slots.every(x => timed(x) && strings(x, ['roleCode', 'requiredCount', 'eligibilityCode']))) &&
    (data.sessions === undefined || Array.isArray(data.sessions) && data.sessions.length <= 20 && data.sessions.every(x => timed(x) && text(x.title) && Array.isArray(x.items) && x.items.length <= 50 && x.items.every(i => i && strings(i, ['id', 'startOffsetMinutes', 'durationMinutes']) && text(i.title) && text(i.description)))) &&
    (data.venues === undefined || Array.isArray(data.venues) && data.venues.length <= 20 && data.venues.every(x => timed(x) && strings(x, ['venueId', 'venueETag', 'capacity', 'requiredCapacity']) && text(x.name) && text(x.address)))
}

File: backend/src/Alife.Application/Events/Services/EventCreationArrangements.cs
using Alife.Application.Common.Interfaces;
using Alife.Application.Common.Models;
using Alife.Application.Events.Dtos;
using Alife.Domain.Entities;
using Alife.Domain.Enums;
using Microsoft.EntityFrameworkCore;

namespace Alife.Application.Events.Services;

public sealed record PreparedEventArrangements(IReadOnlyList<EventServiceSlot> Slots,
    IReadOnlyList<EventSession> Sessions, IReadOnlyList<EventVenue> NewVenues,
    IReadOnlyList<EventVenue> ReservedVenues, IReadOnlyList<EventVenueReservation> Bookings);

public static class EventCreationArrangements
{
    // Repeated arrangements are relative to each materialized occurrence, including DST changes.
    public static async Task<AppResult<PreparedEventArrangements>> PrepareAsync(IAlifeDbContext db,
        GroupEvent groupEvent, Guid actorId, EventPlanProposalDto plan,
        IReadOnlyList<EventOccurrence> occurrences, EventCreationArrangementsRequest request,
        DateTime now, CancellationToken ct, IReadOnlySet<Guid>? replacedReservationIds = null)
    {
        AppResult<PreparedEventArrangements> Invalid(string message) => AppResult<PreparedEventArrangements>.Validation(message);
        bool Enabled(string code) => plan.ModuleDecisions.Any(x => x.ModuleCode == code && x.Status != EventModuleDecisionStatus.Inactive);
        var slots = request.ServiceSlots ?? [];
        var sessions = request.Sessions ?? [];
        var bookings = request.VenueBookings ?? [];
        if (slots.Count > 50 || sessions.Count > 20 || bookings.Count > 20 || sessions.Any(x => x is null || x.Items is null || x.Items.Count > 50))
            return Invalid("Creation arrangements exceed the supported limits (50 slots, 20 sessions, 50 items per session, 20 bookings).");
        if (slots.Count > 0 && !Enabled("SERVICE.ROSTER") || sessions.Count > 0 && !Enabled("PROGRAM.PRODUCTION") || bookings.Count > 0 && !Enabled("PLACE.RESOURCE"))
            return Invalid("Arrangement details require their module to be enabled by the accepted plan.");
        foreach (var slot in slots)
        {
            if (slot is null || string.IsNullOrWhiteSpace(slot.RoleCode) || slot.RoleCode.Length > 100 ||
                slot.RequiredCount is < 1 or > 10000 || !ValidTime(slot.StartOffsetMinutes, slot.EndOffsetMinutes) ||
                string.IsNullOrWhiteSpace(slot.EligibilityCode) || slot.EligibilityCode.Length > 120 ||
                slot.EligibilityCode != "approvedGroupMember" && slot.EligibilityCode != "acceptedEventTeamMember" &&
                !(slot.EligibilityCode.StartsWith("acceptedRole:", StringComparison.Ordinal) && slot.EligibilityCode.Length > 13))
                return Invalid("Each service slot needs a role, valid eligibility, positive whole-number count and valid times.");
        }
        foreach (var session in sessions)
        {
            if (!ValidText(session.Title, 240) || !ValidTime(session.StartOffsetMinutes, session.EndOffsetMinutes))
                return Invalid("Each programme session needs bilingual titles and valid times.");
            if (session.Items.Any(item => item is null || !ValidText(item.Title, 240) ||
                item.Description is not null && !ValidText(item.Description, 2000, false) ||
                item.StartOffsetMinutes < 0 || item.DurationMinutes < 1 ||
                (long)item.StartOffsetMinutes + item.DurationMinutes > session.EndOffsetMinutes - session.StartOffsetMinutes))
                return Invalid("Programme items need bilingual titles and must fit inside their session.");
        }
        if (occurrences.Any(x => x.StartUtc < DateTime.UnixEpoch.AddDays(7) || x.StartUtc.Year > 9900))
            return Invalid("Event dates are outside the supported arrangement range.");

        var preparedSlots = occurrences.SelectMany(occurrence => slots.Select(slot => new EventServiceSlot
        {
            Id = Guid.NewGuid(), OccurrenceId = occurrence.Id, RoleCode = slot.RoleCode.Trim(),
            RequiredCount = slot.RequiredCount, EligibilityCode = slot.EligibilityCode.Trim(),
            StartUtc = occurrence.StartUtc.AddMinutes(slot.StartOffsetMinutes), EndUtc = occurrence.StartUtc.AddMinutes(slot.EndOffsetMinutes),
            CreatedUtc = now, UpdatedUtc = now
        })).ToArray();
        var preparedSessions = occurrences.SelectMany(occurrence => sessions.Select(session =>
        {
            var entity = new EventSession
            {
                Id = Guid.NewGuid(), OccurrenceId = occurrence.Id, TitleEn = session.Title.En.Trim(), TitleZh = session.Title.Zh.Trim(),
                StartUtc = occurrence.StartUtc.AddMinutes(session.StartOffsetMinutes), EndUtc = occurrence.StartUtc.AddMinutes(session.EndOffsetMinutes),
                Status = EventSessionStatus.Draft, CreatedUtc = now, UpdatedUtc = now
            };
            entity.ProgramItems = session.Items.Select((item, index) => new EventProgramItem
            {
                Id = Guid.NewGuid(), SessionId = entity.Id, TitleEn = item.Title.En.Trim(), TitleZh = item.Title.Zh.Trim(),
                DescriptionEn = item.Description?.En.Trim() ?? "", DescriptionZh = item.Description?.Zh.Trim() ?? "",
                StartOffsetMinutes = item.StartOffsetMinutes, DurationMinutes = item.DurationMinutes, SortOrder = (index + 1) * 10,
                CreatedUtc = now, UpdatedUtc = now
            }).ToArray();
            return entity;
        })).ToArray();
        var newVenues = new List<EventVenue>();
        var reservedVenues = new Dictionary<Guid, EventVenue>();
        var reservations = new List<EventVenueReservation>();
        foreach (var booking in bookings)
        {
            if (booking is null || !ValidTime(booking.StartOffsetMinutes, booking.EndOffsetMinutes) || booking.RequiredCapacity is < 1 or > 1000000 ||
                booking.VenueId.HasValue == (booking.NewVenue is not null))
                return Invalid("Each booking needs valid times, capacity and either an existing or new venue.");
            EventVenue venue;
            if (booking.VenueId.HasValue)
            {
                venue = await db.EventVenues.FirstOrDefaultAsync(x => x.Id == booking.VenueId && x.ManagingGroupId == groupEvent.GroupId, ct) ?? null!;
                if (venue is null || !venue.IsActive) return Invalid("The selected venue is unavailable in this group.");
                if (booking.VenueETag != $"\"venue-{venue.ConcurrencyToken:N}\"")
                    return AppResult<PreparedEventArrangements>.PreconditionFailed("The selected venue changed. Refresh venues in Arrangements and review again.");
            }
            else
            {
                var source = booking.NewVenue!;
                if (!ValidText(source.Name, 240) || source.Address is not null && !ValidText(source.Address, 1000, false) ||
                    source.Capacity is < 1 or > 1000000 || !source.IsActive)
                    return Invalid("New venues need bilingual names, a positive capacity and active status.");
                venue = new EventVenue { Id = Guid.NewGuid(), ManagingGroupId = groupEvent.GroupId,
                    NameEn = source.Name.En.Trim(), NameZh = source.Name.Zh.Trim(),
                    AddressEn = source.Address?.En.Trim() ?? "", AddressZh = source.Address?.Zh.Trim() ?? "",
                    Capacity = source.Capacity, IsActive = true, CreatedByMemberId = actorId, CreatedUtc = now, UpdatedUtc = now };
                newVenues.Add(venue);
            }
            if (booking.RequiredCapacity > venue.Capacity) return Invalid("The selected venue capacity is below the requested attendance.");
            reservedVenues[venue.Id] = venue;
            foreach (var occurrence in occurrences)
            {
                var start = occurrence.StartUtc.AddMinutes(booking.StartOffsetMinutes);
                var end = occurrence.StartUtc.AddMinutes(booking.EndOffsetMinutes);
                if (reservations.Any(x => x.VenueId == venue.Id && x.StartUtc < end && start < x.EndUtc) ||
                    await db.EventVenueReservations.AsNoTracking().AnyAsync(x => x.VenueId == venue.Id &&
                        x.Status == EventVenueReservationStatus.Confirmed && x.StartUtc < end && start < x.EndUtc &&
                        (replacedReservationIds == null || !replacedReservationIds.Contains(x.Id)), ct))
                    return AppResult<PreparedEventArrangements>.Conflict("A selected venue conflicts with another booking. Adjust the venue or time in Arrangements.");
                reservations.Add(new EventVenueReservation { Id = Guid.NewGuid(), VenueId = venue.Id,
                    EventId = groupEvent.Id, EventOccurrenceId = occurrence.Id, StartUtc = start, EndUtc = end,
                    RequiredCapacity = booking.RequiredCapacity, Status = EventVenueReservationStatus.Confirmed,
                    ReservedByMemberId = actorId, CreatedUtc = now, UpdatedUtc = now });
            }
        }
        return AppResult<PreparedEventArrangements>.Success(new(preparedSlots, preparedSessions, newVenues, reservedVenues.Values.ToArray(), reservations));
    }

    private static bool ValidTime(int start, int end) => start >= -10080 && end <= 44640 && end > start;
    private static bool ValidText(LocalizedTextDto? value, int limit, bool required = true) => value is not null &&
        value.En is not null && value.Zh is not null && value.En.Length <= limit && value.Zh.Length <= limit &&
        (!required || !string.IsNullOrWhiteSpace(value.En) && !string.IsNullOrWhiteSpace(value.Zh));
}
